src/dune_engine/action_mapper.ml

```diff
@@ -51,9 +51,29 @@ module Make (Src : Action_intf.Ast) (Dst : Action_intf.Ast) = struct
         , f_target ~dir target )
     | No_infer t -> No_infer (f t ~dir)
     | Pipe (outputs, l) -> Pipe (outputs, List.map l ~f:(fun t -> f t ~dir))
-    | Format_dune_file (src, dst) ->
-      Format_dune_file (f_path ~dir src, f_target ~dir dst)
     | Cram script -> Cram (f_path ~dir script)
+    | Extension
+        ( v
+        , { name
+          ; version
+          ; deps
+          ; targets
+          ; action
+          ; how_to_cache
+          ; encode
+          ; simplified
+          } ) ->
+      Extension
+        ( v
+        , { name
+          ; version
+          ; deps = (fun v -> List.map ~f:(f_path ~dir) (deps v))
```

Comment:
I think it's not sufficient to only map `deps` and `targets`.
We also have to map the paths used by the `action` itself for sandboxing to work properly.
(the action needs to somehow know that the paths embedded in `v` are referring to sandbox)
I think a brute-force way to make it work would be an exercise in encoding of higher-kinded types in OCaml module system, which can be fun, but may end up pretty confusing.

I guess one way around it could be to additionally store `f_path : path0 -> path` and `f_target : target0 -> target` functions themselves in the record, a bit similar to what you're doing with `targets` and `deps` already
("path0" and "target0" would refer to the types actually embedded into `v`).

A separate performance concern is that here you we up re-allocating the record that I thought we'd only have to allocate once.
We should probably move the dynamic parts out of the record, into the (smaller) enclosing tuple, and pass the big record unmodified.

Reply:
I'm not entirely sure there's a great story for this either way.

I'll have to look more closely at sandboxing in general, but could this be addressed by only exposing `Extension` as a fully-expanded action? Right now, extensions will need to transform their own arguments into `Dpath.t` (etc), so maybe this could be done there?

Reply:
Yeah, if you find a way to avoid going via action_mapper entirely that could also work. Although for sandboxing I imagine something like that will be needed.

Reply:
Hmm, yeah, I think there's no way around it.

Could we just include `map_embedded : f_path:(...) -> f_target:(...) -> ... -> 'a -> 'a` as a field of the record? I think that would solve our problem.

Reply:
With regards to `action` itself, could we solve this with the `register` api? Then we could ensure that the closure doesn't capture any paths, and that those paths should be in `v` itself?

Reply:
> Could we just include `map_embedded : ... -> 'a -> 'a`

Conceptually that's exactly what we need, but the type will grow much hairier because `f_path` and `f_target` change the types, so `'a` will also have to change (it won't be `'a -> 'a`).

> `action` itself

I meant the paths in `v` that action uses, not the ones embedded in the closure. Having anything embedded in the closure would indeed be a mistake, and the register API seems like a fine solution, but it's somewhat orthogonal to this conversation.

Reply:
> the type will grow much hairier because `f_path` and `f_target` change the types

I don't think we really need to worry about this. As extensions should always be constructed fully-expanded, we only need to worry about a few cases, namely

- Sandboxing
- Relativisation
- Unresolved -> Resolved

The types only actually change when mapping to `Relativise`, but it doesn't matter a whole lot, because that's only used in `for_shell`, which is in turn only used for pretty printing, digests, stamping, etc, so it doesn't really matter.

> I meant the paths in `v` that action uses, not the ones embedded in the closure.

I guess I'm not seeing why `action` needs to be aware of whether things are sandboxed, then.

Reply:
Regarding the mapper, one (ad-hoc) idea may be to just add more elements to the record for each of the above three cases. We can use `map_one_step` to make sure that this works properly.

Reply:
Actually, I don't even think we have to worry about path resolution, since that mainly matters when expanding.

Reply:
Okay, so we might be able to solve both these issues (the mapper problem and the record allocation) with one stroke:

- In `action_mapper` and `action_unexpanded`, we raise a code error if an `Expansion` variant is attempted to be expanded. This should always be a logic error on our side, because there are only two `Action_mapper`s outgoing from `Action.Ast`, both of which will be dealt with separately.
- The two outgoing mappers are for sandboxing and `Relativise`. We can then use `map_one_step` to intercept the `Extension` variant and instead do something else.
  - For the first, we can provide the field `map_same_type : f_path:(Path.t -> Path.t) -> ... -> 'a -> 'a`. This is well-typed, because the sandboxed and non-sandboxed asts are the same.
  - For `Relativise`, it's safe to just ignore the transformation and pass the variant unchanged (maybe we still need to modify the `deps` and `targets` fields, as they're used in digests), since we ignore everything except the deps/targets/`simplify` field anyway.